pandas/_libs/groupby.pyx

```diff
@@ -117,6 +121,97 @@ def group_last_object(ndarray[object, ndim=2] out,
             else:
                 out[i, j] = resx[i, j]
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def group_rank_object(ndarray[float64_t, ndim=2] out,
+                      ndarray[object, ndim=2] values,
+                      ndarray[int64_t] labels,
+                      bint is_datetimelike, object ties_method,
+                      bint ascending, bint pct, object na_option):
```

Comment:
ideally could incorporate this in the templated one. though I would just raise on object dtype. seems odd to support this (as its a very specific encoding, meaning lexical).

Reply:
Simply raising on an object type would be easy. That said, wouldn't that then make this inconsistent with the ``nth`` / ``first`` / ``last`` functions? 

It seems strange to me if we end up allowing users to pick the ``nth`` object from a group of objects but do not allow them to rank those same objects

Reply:
Can ignore this comment. Was thinking at the time that `nth` had to do with the value of an item but it has more to do with the position. I'll push another PR raising for objects. Will have to communicate as a breaking change in case anyone had previously been using that behavior

Reply:
@jreback just scoping this out I think this is a fairly significant change. The main problem I see is that the current object ranking implementation is used by Categoricals as well. I imagine we'd have to then update it so that ordered Categoricals can use rank but non-ordered Categoricals cannot. Additionally would have to update the code across Series, Frame and GroupBy objects to ensure those handle consistently.

I think that strays a little too far from the original goal here of optimizing performance and would rather handle in a separate change. Let me know if you disagree otherwise happy to open that one.

Reply:
absolutely. let's keep this to one item for this PR (you can open a separate issue for other)

Reply:
Opened #19560 for this issue. Could use some further discussion.

Otherwise if this round of tests pass I plan to update whatsnew and wrap this one up. If there's anything outstanding please let me know

Reply:
I would like to raise on groupby_object_rank rather than implement it here. let's keep that for another discussion.

Reply:
I would simply not define it at all. This will be caught at a higher level (IOW it will try to compose the function name it won't exist), I think the error is ok.